docs(serviceprovider): add quality criteria first draft (#105)
openmcp-project/docs: docs/developers/serviceprovider/07-quality-criteria.mdx

---
sidebar_position: 7
id: quality-criteria
---

# Quality Criteria

This page defines the quality criteria a `ServiceProvider` should aim for to be ready for use in different kinds of OpenControlPlane landscapes, from early-stage experiments to production environments with corporate compliance requirements.

> Use this page as a starting point: it tells you which criteria to implement and helps you understand what maturity level your service provider is at. Once you have assessed your own provider against the criteria, a sig-extensibility approver can review your assessment and confirm which tier it belongs to. No further process is required.

Reviewer: Imo it is just a self assessment. I don't see us doing any governance or quality assessments on service providers we do not maintain.

```suggestion
Use this page as a starting point: it tells you which criteria to implement and helps you understand what maturity level your service provider is at.
```

| <img src="https://img.shields.io/badge/Quality-Experimental-e69138?style=flat-square&labelColor=555" alt="Quality: Experimental" /> | Proof-of-concept, evaluation, contributor hacking                                     | Anything you don't want to clean up by hand if it breaks       |
| <img src="https://img.shields.io/badge/Quality-Incubating-3d9970?style=flat-square&labelColor=555" alt="Quality: Incubating" />     | Dev/test landscapes, internal platforms, situations where rough edges are tolerable   | Regulated or production environments without additional review |
| <img src="https://img.shields.io/badge/Quality-Stable-1e8f95?style=flat-square&labelColor=555" alt="Quality: Stable" />             | Production landscapes, customer-facing platforms, corporate / air-gapped environments |                                                                |

## The quality criteria

### 1. Deletion behaviour

> When a user triggers an uninstall of a domain service through the `ServiceProviderAPI`, the service provider must not blindly remove the domain service and its CRDs from the managed `ControlPlane`. Domain services like Crossplane, Flux, or ESO install CRDs into the `ControlPlane`, and users may have created instances of those CRDs. Deleting the CRDs while custom resources still exist would leave those resources orphaned in the cluster or cause managed resources in external systems to escape the platform's control.

Reviewer: nit: I think we can generally drop the `domain service` wording in this document. It is required in the design docs and partly at some points in the development guides to clearly distinguish between service provider domain and domain service (domain) but imo this doc would benefit from just speaking about services, e.g.

```suggestion
When a user requests to uninstall a service through the `ServiceProviderAPI`, the service provider must not blindly remove the service and its CRDs from the managed `ControlPlane`. Services like Crossplane, Flux, or ESO install CRDs into the `ControlPlane`, and users may have created instances of those CRDs. Deleting the CRDs while custom resources still exist would leave those resources orphaned in the cluster or cause managed resources in external systems to escape the platform's control.
```

### 8. Ownership and maintenance documentation

A `MAINTAINERS.md` or `CODEOWNERS` file names responsible humans or teams. Contribution and issue reporting instructions are documented in `CONTRIBUTING.md` or the project README.

## Declaring a tier in your service-provider repository

Each service-provider repository carries a **Quality Criteria block** in its README, placed directly below the title and badges and above the "About this project" section. It has three parts.

### 1. Tier badge

A badge showing the tier you claim. Copy the snippet for your tier from the [Badge snippets](#badge-snippets) section below and paste it into your README.

### 2. Self-assessment table

A table showing the status of every criterion. Three states:

- ✅ **Met** — the criterion is fully satisfied.
- ❌ **Not met** — the criterion is not satisfied.
- ⚠️ **Partial** — some aspects are met, with known gaps documented (link to a tracking issue if available).

> Fill this in yourself based on your own assessment. A sig-extensibility approver can then review it and confirm the tier. No further process is required.

Reviewer: same as before, no governance/assessment on service providers that we don't maintain.

```suggestion
Fill this in yourself based on your own assessment.
```

---
sidebar_position: 7
id: quality-standards
---

# Quality Standards

This page defines the quality standards a `ServiceProvider` should meet to be considered ready for use in different kinds of OpenControlPlane landscapes, from early-stage experiments to production environments with corporate compliance requirements.

The standard exists for two audiences:

1. **Service provider developers** read it as a checklist. It tells you what you need to implement and which tier you can claim.
2. **Platform owners** read it to evaluate whether a service provider is mature enough to install in their landscape.

The standard is **tooling-agnostic**. It describes what a service provider must *produce* (its artifacts) and how it must *behave* (at runtime), not how it is built. Service providers in any GitHub organisation can adhere to it, regardless of the build system or CI they use.

> ## Tiers at a glance

Reviewer: `Community` (governance/ownership) feels like a different dimension vs `Stable` and `Experimental` (technical maturity).

ownership dimension could be `Official` (provided by the core team) and `Community` and maybe something in between like `Verified` which aligns with your suggestion of core maintainers reviewing external community repos.

When I look at the overview table that is provided at the end of this document, maybe rename the tiers to `Experimental`, `Stabilizing` and `Stable/Production-ready`?

Author: I have another proposal: How about `Experimental` -> `Incubating` -> `Stable`? Right now, I just outline that these tiers exist briefly and they will get more meaning when a Registry/Marketplace concept will evolve. What do you think?

- Resources expose typed Conditions and a Phase (`Ready` / `Progressing` / `Terminating`).
- Documentation includes an install guide and a configuration reference.
- A `MAINTAINERS.md` or `CODEOWNERS` file identifies who responds to issues.
- Unit tests cover the reconciler.

### Stable

<img src="https://img.shields.io/badge/Quality-Stable-1e8f95?style=flat-square&labelColor=555" alt="Quality: Stable" />

Production-ready. Fit for use in corporate, regulated, and air-gapped environments.

**Guarantees** (in addition to Community):

- Air-gapped: provider supports custom CA bundles.
- Security hardened: container runs as non-root, with a read-only root filesystem and dropped Linux capabilities; RBAC follows least-privilege.
- API is at least `v1beta1`; breaking changes require a deprecation cycle.
- End-to-end tests run on every release against a real cluster, e.g. using [openmcp-testing](https://github.com/openmcp-project/openmcp-testing).
- Documentation includes a troubleshooting section.
- Deletion is well-defined: before uninstalling a domain service, the provider checks for existing custom resources belonging to it and blocks removal if any are found.

> ## The quality criteria

Reviewer: Please add a section regarding required annotations where we point to our controller guidelines https://open-control-plane.io/developers/general#operation-annotations

Author: Addressed with [`56b8569` (this PR)](https://github.com/openmcp-project/docs/pull/105/commits/56b856934bbd0d88736b981e24442aac3f7607f1) and ff.
